environments/centos7-base/main.py

```diff
@@ -0,0 +1,204 @@
+#!/usr/bin/env python3
+
+import os
+import sys
+import stat
+import json
+import urllib2
+
+from datetime import datetime
+from subprocess import call
+from string import Template
+
+def error(message):
+    log(Template('ERR: $message').substitute(message=message))
+    sys.stdout.flush()
+
+def warn(message):
+    log(Template('WARN: $message').substitute(message=message))
+    sys.stdout.flush()
+
+def log(message):
+    print(Template('[main] $message').substitute(message=message))
+    sys.stdout.flush()
+
+# This function will attempt to read any results file, add our metadata to it,
+# write it to disk, send everything to S3, and exit the program.
+#
+# If succeeded is false, we'll assume things failed badly enough that we shouldn't
+# even try to read /grade/results/results.json.
+def finish(succeeded, info):
+    if info['job_id'] is None:
+        # Someone screwed up bad; without a job ID, we can't upload archives to
+        # S3 or notify the webhook that we've failed
+        error('job_id was not specified, so it\'s impossible to upload an archive')
+        sys.exit(1)
+
+    data = None
+    if succeeded and os.path.isfile('/grade/results/results.json'):
```

Comment:
I'm thinking about the various ways that something should go wrong, how those errors should be reported, and who they should be reported to. It seems to me we have three different types of errors: (1) system errors, like a missing `grading_id` or a failed write to an S3 bucket, (2) autograder errors, like invalid JSON output by the autograder, and (3) errors in student code, like a missing file. Similarly, we have three different types of people to report to: (A) system admins, (B) course instructors, (C) the student. I think that all error types need to be reported to C, types 1 and 2 need to be reported to B, and only type 1 errors should be reported to A. However, aggregate statistics on type 2 errors might be wanted by A to see which courses are having issues. Initially it's more important that we record error data, and we can worry about reporting interfaces later.

It's possible in the error classification scheme above that type 3 errors are not really "errors", but are just handled by the autograder in the regular way with feedback. I'm not sure about this, however. For example, if there is a missing file in the student submission, is this a different kind of error than a compile error, or a failing testcase? Maybe we want to just handle all of them the same.

When we are reporting errors, it's important to be clear about who's "fault" the error is. Maybe we can report these as "system error", "autograder error", etc. We also shouldn't leak private information, so course instructors shouldn't see details of our AWS infrastructure, and students shouldn't see autograder details.

One useful technique that PL uses for errors already is to generate a random unique errorId (see https://github.com/PrairieLearn/PrairieLearn/blob/master/pages/error/error.js#L14). This is reported to the student without additional details, but then the log files contain the same errorId with full debug output (stack traces, internal variable states, etc). This way the sysadmins can get a bug report saying "I did XXX and it said errorId: 4K8XU81", and then it's easy to go look up the detailed information in the logs.

I'm thinking that when we have an error, we want to:

1. Generate a random errorId. We could actually just make these UUIDs.

2. Set the errorType to either "system", "course", or "student" (maybe we only just use the first two?).

3. Set systemErrorMessage, systemErrorData, courseErrorMessage, courseErrorData, studentErrorMessage, studentErrorData. These might all be the same, or they might contain variable levels of detail.

4. Record all of the above along with an "error: true" flag in the job output data.

Example 1: results JSON doesn't parse:

* `errorType = 'course'`
* `systemErrorMessage = courseErrorMessage = 'Unable to parse result JSON'`
* `systemErrorData = courseErrorData = {rawData: data, parseError: msg}'
* `studentErrorMessage = 'Autograder error'`
* `studentErrorData = null`

Example 2: unable to write to S3:

* `errorType = 'system'`
* `systemErrorMessage = 'Error writing to S3 bucket'`
* `systemErrorData = {bucket: 'XXXX', errorMsg: msg}`
* `courseErrorMessage = studentErrorMessage = 'System error'`
* `courseErrorData = studentErrorData = null`

In all cases we report the errorUuid to everyone (course staff, students, etc), and we only report the appropriate error message and error data.

To support this error reporting infrastructure we should eventually make a new DB table called `errors` to track all errors. We don't need to do this as part of the autograder branch, however. This will also address #592.